### PR TITLE
fix: memory leak in menubar

### LINKED ISSUE #198051: memory leak in menubar
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.84.2
- OS Version: Ubuntu 23.04

Steps to Reproduce:

1. Open a VSCode Window
2. In the menu at the top, click on "File" which opens the menu
3. In the menu at the top, click on "File" again which closes the menu
4. A Menu instance has been added, but not been removed

### Test Script

```sh
git clone git@github.com:SimonSiefke/vscode-memory-leak-finder.git &&
cd vscode-memory-leak-finder &&
git checkout v5.22.2 &&
npm ci &&
node packages/cli/bin/test.js --cwd packages/e2e  --check-leaks --measure instance-counts-with-source-map --only title-bar &&
cat .vscode-memory-leak-finder-results/instance-counts-with-source-map/title-bar-menu-toggle.json
```

### Output

```json
{
  "instanceCounts": {
    "before": [
      {
        "name": "Menu",
        "count": 1
      }
    ],
    "after": [
      {
        "name": "Menu",
        "count": 2
      }
    ]
  },
  "isLeak": true
}
```

### PR DESCRIPTION
**Pull request by SimonSiefke**
Fixes #198051

### Output with `--measure instance-counts`

```json
{
  "instanceCounts": {
    "before": [
      {
        "name": "Menu",
        "count": 1
      }
    ],
    "after": [
      {
        "name": "Menu",
        "count": 1
      }
    ]
  },
  "isLeak": false
}
```

### CODE REVIEW

File: src/vs/base/browser/ui/menu/menubar.ts [hunk 1]
```diff
@@ -87,6 +87,8 @@ export class MenuBar extends Disposable {
 	private numMenusShown: number = 0;
 	private overflowLayoutScheduled: IDisposable | undefined = undefined;
 
+	private menuDisposables = new DisposableStore();
```
Comment (bpasero) [inline src/vs/base/browser/ui/menu/menubar.ts:90]:
```suggestion
	private readonly menuDisposables = this._register(new DisposableStore());
```

File: src/vs/base/browser/ui/menu/menubar.ts [hunk 2]
```diff
@@ -985,6 +987,8 @@ export class MenuBar extends Disposable {
 
 			this.focusedMenu = { index: this.focusedMenu.index };
 		}
+		this.menuDisposables.dispose();
+		this.menuDisposables = new DisposableStore();
 	}
```
Comment (bpasero) [inline src/vs/base/browser/ui/menu/menubar.ts:992]:
```suggestion
		this.menuDisposables.clear();
	}
```

File: src/vs/base/browser/ui/menu/menubar.ts [hunk 3]
```diff
@@ -1025,9 +1029,9 @@ export class MenuBar extends Disposable {
 			useEventAsContext: true
 		};
 
-		const menuWidget = this._register(new Menu(menuHolder, customMenu.actions, menuOptions, this.menuStyle));
-
-		this._register(menuWidget.onDidCancel(() => {
+		const menuWidget = new Menu(menuHolder, customMenu.actions, menuOptions, this.menuStyle);
+		this.menuDisposables.add(menuWidget);
+		this.menuDisposables.add(menuWidget.onDidCancel(() => {
```
Comment (bpasero) [inline src/vs/base/browser/ui/menu/menubar.ts:1034]:
```suggestion
		const menuWidget = this.menuDisposables.add(new Menu(menuHolder, customMenu.actions, menuOptions, this.menuStyle));
		this.menuDisposables.add(menuWidget.onDidCancel(() => {
```

**Review by bpasero — DISMISSED** [review]

**Comment by bpasero** [comment]
I think from here on I have to defer to Steven to review because I am not the original author of the code. Its not clear to me all the paths that lead to re-create or dispose of the custom menu...

**Review by sbatten — APPROVED** [review]

**Review by rzhao271 — APPROVED** [review]

**Pull request merged**